Use hatch environments and scripts for styling, tests, docs & CI

BSchilperoort opened the pull request:
Hatch scripts are set up for;

- Checking code style, and formatting the code
- Running the tests, running the tests+coverage (for codecov)
- Generating the docs

In Github Actions, the scripts now use these hatch commands as well.

contribute.rst has been updated with these new commands (making it a lot more succinct and easy for developers).

ReadTheDocs sadly still depends on the pip install, and we're not able to configure the build script there 👎

BSchilperoort commented:
@Peter9192 I have now defined hatch envs for every group of scripts. Such as an environment `test` with the scripts `check` and `cov`. However alternatively, we could put the test and styling script (and dependencies) under default. 

E.g.
```toml
[tool.hatch.envs.default]
features = ["dev",]

[tool.hatch.envs.default.scripts]
test = ["pytest",]
style = [
  "flake8 .",
  "black --check --diff .",
  "isort --check-only --diff .",
]
fmt = ["isort .", "black .", "fmt",]
```

This would simplify things, as anyone developing would need to have those tools anyway. Both tools also need to know about the package+dependencies (can't be detached).

The docs build environment could stay separate however.

What's your opinion?

Peter9192 commented:
> What's your opinion?

I think I agree, as long as the default env doesn't grow too complex.

BSchilperoort commented:
@Peter9192 I have implemented what I suggested.

Additionally, I modified the instructions to suggest the use of `hatch shell`, instead of creating a virtual environment.

Peter9192 reviewed (approved):
Tried everything, works like a charm!

BSchilperoort commented on docs/source/contribute.rst:
```diff
@@ -15,50 +15,42 @@ using and the version of the cdsapi library installed.
-
-Before pushing a new addition, some tools are required to confirm that the code
-is up to standard.
+   hatch shell  # Or: python3 -m hatch shell
```
Yes, nice! Thanks for pointing that out.
